runtime-sdk/modules/evm/src/precompile/erc20.rs

```diff
@@ -0,0 +1,556 @@
+/// token-specific configuration to the static contract framework.
+pub trait Erc20Token {
+    const GAS_COSTS: TokenOperationCosts;
+
+    const NAME: &str;
+    const SYMBOL: &str;
+    const DECIMALS: u8;
+
+    fn address() -> H160;
+
+    fn total_supply() -> Result<u128, Error>;
+    fn balance_of(account: &H160) -> Result<u128, Error>;
+    fn transfer(sender: &H160, recipient: &H160, amount: u128) -> Result<bool, Error>;
+    fn transfer_from(
+        owner: &H160,
+        spender: &H160,
+        recipient: &H160,
+        amount: u128,
+    ) -> Result<bool, Error>;
+    fn approve(owner: &H160, spender: &H160, amount: u128) -> Result<bool, Error>;
+    fn allowance(owner: &H160, spender: &H160) -> Result<u128, Error>;
+    fn mint(caller: &H160, to: &H160, amount: u128) -> Result<(), Error>;
+    fn burn(caller: &H160, from: &H160, amount: u128) -> Result<(), Error>;
+}
+
+/// Helper trait for simplifying implementations for tokens backed by the SDK's
+/// Accounts module.
+///
+/// The trait presents a much narrower interface than the full [`Erc20Token`]
+/// trait and there is a blanket [`Erc20Token`] impl for any struct that
+/// implements `AccountToken`.
+pub trait AccountToken {
+    type Accounts: AccountsAPI;
+
+    const ADDRESS: H160;
+
+    const GAS_COSTS: TokenOperationCosts;
+
+    const NAME: &str;
+    const SYMBOL: &str;
+    const DECIMALS: u8;
+
+    fn denomination() -> token::Denomination;
+
+    fn is_minting_allowed(caller: &H160, address: &H160) -> Result<bool, Error>;
+    fn is_burning_allowed(caller: &H160, address: &H160) -> Result<bool, Error>;
+}
+
+fn with_allowance<T, F>(owner: &H160, spender: &H160, f: F) -> Result<u128, Error>
+where
+    T: AccountToken,
+    F: FnOnce(Option<u128>) -> Result<u128, Error>,
+{
+    let owner = owner.as_bytes();
+    let spender = spender.as_bytes();
+    CurrentState::with_store(|store| {
+        let store = storage::PrefixStore::new(store, &MODULE_NAME);
+        let allowances = storage::PrefixStore::new(store, &STATE_ALLOWANCES);
+        let denomination = T::denomination().to_string();
+        let allowances = storage::PrefixStore::new(allowances, &denomination);
+        let mut owner = storage::TypedStore::new(storage::PrefixStore::new(allowances, &owner));
+        let old_allowance = owner.get(spender);
+        let new_allowance = f(old_allowance)?;
+        if new_allowance == 0 {
+            owner.remove(spender);
+        } else {
+            owner.insert(spender, new_allowance);
+        }
+        Ok(old_allowance.unwrap_or_default())
+    })
+}
+
+impl<T: AccountToken> Erc20Token for T {
+    const GAS_COSTS: TokenOperationCosts = T::GAS_COSTS;
+
+    const NAME: &str = T::NAME;
+    const SYMBOL: &str = T::SYMBOL;
+    const DECIMALS: u8 = T::DECIMALS;
+
+    fn address() -> H160 {
+        Self::ADDRESS
+    }
+
+    fn total_supply() -> Result<u128, Error> {
+        Ok(T::Accounts::get_total_supplies()?
```

Comment:
Add `Accounts::get_total_supply(denomination)` to avoid first iterating over all total supplies for no good reason. In storage, each supply is its own storage location so this should be straightforward.